Add regions parameter in aws module config

kaiyan-sheng opened the pull request:
This PR is to add `regions` in config for aws module. With `regions`, we give users the ability to specify what are the regions they want to collect cloudwatch metrics from for different AWS services. `regions` is optional, when there is no `regions` in the config, the code will get a full list of AWS regions and loop through all of them. 

Please see https://github.com/elastic/beats/issues/11932 for more details.

kaiyan-sheng commented:
One concern I have is, will the change consider a breaking change? If so, we can keep the original config `default_region` and add `regions` on top of that.

exekias commented on x-pack/metricbeat/module/aws/_meta/config.yml:
```diff
@@ -1,18 +1,27 @@
+  access_key_id: '${AWS_ACCESS_KEY_ID:""}'
+  secret_access_key: '${AWS_SECRET_ACCESS_KEY:""}'
+  session_token: '${AWS_SESSION_TOKEN:""}'
+  regions:
```
This should default to query from all regions, I suggest to keep `regions` setting here but commented out

kaiyan-sheng replied:
Good point, I just updated the config.yml with regions commented.

exekias commented on x-pack/metricbeat/module/aws/aws.go:
```diff
@@ -62,25 +62,19 @@ func NewMetricSet(base mb.BaseMetricSet) (*MetricSet, error) {
-	if err != nil {
-		return nil, err
-	}
+	awsConfig.Region = "us-east-1"
```
Is there any chance the user may need to configure this? I wonder if it would be better to use `metricSet.RegionsList[0]`

exekias commented:
This is a breaking change, IMHO it should be ok because we offer a new way to configure it. In any case it should be also added to the breaking changes list in the changelog

jsoriano commented:
> This is a breaking change, IMHO it should be ok because we offer a new way to configure it. In any case it should be also added to the breaking changes list in the changelog

Could we keep the old option for backwards compatibility as a synonym of setting regions to a single value?

exekias commented:
@kaiyan-sheng to comment here, but If I understood this correctly, `default_region` was something different from setting a single region. Default region is only used to get the list of regions, we then iterate over them to fetch metrics.

It's more about initialization. I guess it's not a very important setting, although it may be needed in some cases, hence the fallback to take the first region from the list?

kaiyan-sheng commented:
> @kaiyan-sheng to comment here, but If I understood this correctly, `default_region` was something different from setting a single region. Default region is only used to get the list of regions, we then iterate over them to fetch metrics.
> 
> It's more about initialization. I guess it's not a very important setting, although it may be needed in some cases, hence the fallback to take the first region from the list?

@exekias Yes, `default_region` is an optional parameter in config currently and it's only for getting the full list of regions from AWS. It just need _a_ value to start the first AWS query and it doesn't matter which region it's default to. But we do need a default value before `DescribeRegions` API call can be made.

kaiyan-sheng commented:
> > This is a breaking change, IMHO it should be ok because we offer a new way to configure it. In any case it should be also added to the breaking changes list in the changelog
> 
> Could we keep the old option for backwards compatibility as a synonym of setting regions to a single value?

@jsoriano Do you mean keep `default_region` in the config and if user set `default_region`, then only collect metrics from that `default_region`? That's also a changing behavior from previous version hmmm

kaiyan-sheng commented:
jenkins, test this please

jsoriano commented:
> > > This is a breaking change, IMHO it should be ok because we offer a new way to configure it. In any case it should be also added to the breaking changes list in the changelog
> > 
> > 
> > Could we keep the old option for backwards compatibility as a synonym of setting regions to a single value?
> 
> @jsoriano Do you mean keep `default_region` in the config and if user set `default_region`, then only collect metrics from that `default_region`? That's also a changing behavior from previous version hmmm

I misinterpreted `default_region`. I wanted to ask if keeping both options is a possibility, so we don't introduce a breaking change.
Now, understanding better `default_region`, we may still want to keep it to select a region to do this first API call.

kaiyan-sheng commented:
@exekias @jsoriano I added `default_region` back to the config and moved changelog from `breaking changes` session to `added` session. Originally I was thinking to remove `default_region` completely in the future release. But looking at https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-configure.html, it might be good to keep `default_region` just to be consistent with aws cli configuration example.
